app/guard/guard.go

```diff
@@ -0,0 +1,256 @@
+			Aliases: []string{"s"},
+			Usage:   "Setup the configuration for the server (interactive)",
+			Action:  setpass,
+		},
+		{
+			Name:      "setup",
+			Aliases:   []string{"su"},
+			Usage:     "Saves the cothority group-toml to the configuration",
+			ArgsUsage: "Give group definition",
+			Action:    setup,
+		},
+		{
+			Name:    "recover",
+			Aliases: []string{"r"},
+			Usage:   "Gets the password back from the guards",
+			Action:  get,
+		},
+	}
+	app.Before = func(c *cli.Context) error {
+		b, err := ioutil.ReadFile("config.bin")
+		if os.IsNotExist(err) {
+			return nil
+		}
+		log.ErrFatal(err, "The config.bin file threw an error")
+		_, msg, err := network.UnmarshalRegistered(b)
+		log.ErrFatal(err, "UnmarshalRegistered messeed up")
+		var ok bool
+		db, ok = msg.(*Database)
+		if !ok {
+			log.Fatal("We are not OK")
+		}
+		return nil
+	}
+	app.Run(os.Args)
+}
+
+func xor(x []byte, y []byte) []byte {
+	z := make([]byte, len(x))
+
+	for i := 0; i < len(x); i++ {
+		z[i] = x[i] ^ y[i]
+	}
+	return z
+}
+
+// readGroup takes a toml file name and reads the file, returning the entities within
+func readGroup(tomlFileName string) (*sda.Roster, error) {
+	log.Print("Reading From File")
+	f, err := os.Open(tomlFileName)
+	if err != nil {
+		return nil, err
+	}
+	el, err := config.ReadGroupToml(f)
+	if err != nil {
+		return nil, err
+	}
+	if len(el.List) <= 0 {
+		return nil, errors.New("Empty or invalid group file:" +
+			tomlFileName)
+	}
+	log.Lvl3(el)
+	return el, err
+}
+
+func set(c *cli.Context, UID []byte, epoch []byte, password string, userdata []byte) {
+	k := make([]byte, 32)
+	rand.Read(k)
+	//pwhash is the password hash that will be sent to the guard servers
+	pwhash := abstract.Sum(network.Suite, []byte(password), mastersalt)
+	//secretkeys is the Shamir Secret share of the keys
+	secretkeys := s.Create(2, len(db.Cothority.List), string(k))
+	//salty is a list of salts that will be used to encrypt the pwhash before sending it to the guard, derived from a single
+	//master salt
+	salty := saltgen(mastersalt, len(db.Cothority.List))
+	responses := make([][]byte, len(db.Cothority.List))
+	keys := make([][]byte, len(db.Cothority.List))
+	for i := 0; i < len(db.Cothority.List); i++ {
+		cl := guard.NewClient()
+		sendy := abstract.Sum(network.Suite, pwhash, salty[i])
+		rep, err := cl.GetGuard(db.Cothority.List[i], UID, epoch, sendy)
+		log.ErrFatal(err)
+		responses[i] = rep.Msg
+		keys[i] = xor([]byte(secretkeys[i]), responses[i])
+	}
+	//This is the code that seals the user data using the master key and saves it to the db
+	block, _ := aes.NewCipher(k)
+	aesgcm, _ := cipher.NewGCM(block)
+	ciphertext := aesgcm.Seal(nil, mastersalt, userdata, nil)
+	db.Users = append(db.Users, User{UID, mastersalt, keys, ciphertext})
+}
+
+// saltgen is a function that generates all the keys and salts given a length and an initial salt
+func saltgen(salt []byte, count int) [][]byte {
+	salts := make([][]byte, count)
+	tlas := salt
```

Comment:
Why don't you use the salt here? `tlas := salt` doesn't make a copy, if that is your intent:

https://play.golang.org/p/XJ8HsoXSnQ

Perhaps

```
salt = abstract.Sum(network.Suite, salt)
```

is what you need here?